hail/src/main/scala/is/hail/expr/ir/functions/DictFunctions.scala

```diff
@@ -5,30 +5,41 @@ import is.hail.expr.types
 import is.hail.expr.types._
 
 object DictFunctions extends RegistryFunctions {
+  def contains(dict: IR, key: IR) = {
+    val n = Ref(genUID(), TInt32())
+    val i = Ref(genUID(), TInt32())
 
-  def contains(dict: IR, key: IR) =
     If(IsNA(dict),
       NA(TBoolean()),
-      ArrayLen(ToArray(dict)).cne(0) &&
-        ApplyComparisonOp(
-          EQWithNA(key.typ),
-          GetField(ArrayRef(ToArray(dict), LowerBoundOnOrderedCollection(dict, key, onKey=true)), "key"),
-          key))
-
+      Let(n.name,
+        ArrayLen(ToArray(dict)),
+        n.cne(0) && Let(i.name,
```

Comment:
You can actually remove the `n.cne(0)` check now; if the dict is empty, then LowerBound must return 0, and you'll hit the `i == n` case below.